readthedocs/proxito/middleware.py

```diff
@@ -305,6 +306,32 @@ def add_hosting_integrations_headers(self, request, response):
             if addons:
                 response["X-RTD-Hosting-Integrations"] = "true"
 
+    def add_cors_headers(self, request, response):
+        """
+        Add CORS headers only on PUBLIC versions.
+
+        DocDiff addons requires making a request from
+        ``RTD_EXTERNAL_VERSION_DOMAIN`` to ``PUBLIC_DOMAIN`` to be able to
+        compare both DOMs and show the visual differences.
+
+        This request needs ``Access-Control-Allow-Origin`` HTTP headers to be
+        accepted by browsers. However, we cannot expose these headers for
+        documentation that's not PUBLIC.
+        """
+        project_slug = getattr(request, "path_project_slug", "")
+        version_slug = getattr(request, "path_version_slug", "")
+
+        if project_slug and version_slug:
+            allow_cors = Version.objects.filter(
+                project__slug=project_slug,
+                slug=version_slug,
+                privacy_level=PUBLIC,
+            ).exists()
+            if allow_cors:
+                response.headers["Access-Control-Allow-Origin"] = "*.readthedocs.build"
```

Comment:
@ericholscher I don't think you can use a wildcard within the origin, see https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin#directives.

> Only a single origin can be specified. If the server supports clients from multiple origins, it must return the origin for the specific client making the request.

This type of wildcard needs to be implemented in the server, something like:

```python
if is_valid_origing(origin):
    response.headers["Access-Control-Allow-Origin"] = origin
```

Reply:
I think having this to `*` is fine for .org (github and gitlab pages have it to this value, for public pages at least). Maybe we can just start by allowing CORS on .org only.